File: 백준/Silver/1149. RGB거리/RGB거리.java

import java.io.*;
import java.util.*;

public class Main {
    static int RED = 0;
    static int GREEN = 1;
    static int BLUE = 2;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());

        int[][] cost = new int[N][3];
        int[][] dp = new int[N][3];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 3; j++) {
                cost[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        dp[0][RED] = cost[0][RED];
        dp[0][GREEN] = cost[0][GREEN];
        dp[0][BLUE] = cost[0][BLUE];

        for (int i = 1; i < N; i++) {
            dp[i][RED] = Math.min(dp[i - 1][GREEN], dp[i - 1][BLUE]) + cost[i][RED];
            dp[i][GREEN] = Math.min(dp[i - 1][RED], dp[i - 1][BLUE]) + cost[i][GREEN];
            dp[i][BLUE] = Math.min(dp[i - 1][RED], dp[i - 1][GREEN]) + cost[i][BLUE];
        }

        System.out.println(Math.min(Math.min(dp[N - 1][RED], dp[N - 1][GREEN]), dp[N - 1][BLUE]));
    }
}

File: 백준/Silver/1149. RGB거리/RGB거리.py
import collections
import heapq
import sys

input = sys.stdin.readline

n = int(input())
RED = 0
GREEN = 1
BLUE = 2

board = [list(map(int, input().split())) for _ in range(n)]
dp = [[0 for _ in range(3)] for _ in range(n)]

dp[0][RED] = board[0][RED]
dp[0][GREEN] = board[0][GREEN]
dp[0][BLUE] = board[0][BLUE]

for i in range(1, n):
    dp[i][RED] = min(dp[i - 1][GREEN], dp[i - 1][BLUE]) + board[i][RED]
    dp[i][GREEN] = min(dp[i - 1][RED], dp[i - 1][BLUE]) + board[i][GREEN]
    dp[i][BLUE] = min(dp[i - 1][RED], dp[i - 1][GREEN]) + board[i][BLUE]

print(min(dp[n - 1][RED], dp[n - 1][GREEN], dp[n - 1][BLUE]))
